substrate-node/pallets/pallet-smart-contract/src/types.rs

```diff
@@ -286,3 +409,84 @@ pub enum ServiceContractState {
     AgreementReady,
     ApprovedByBoth,
 }
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, scale_info::TypeInfo)]
+pub struct ContractIdProvides<T: Config + Send + Sync + scale_info::TypeInfo>(PhantomData<T>)
+where
+    <T as frame_system::Config>::RuntimeCall: IsSubType<Call<T>>;
+
+impl<T: Config + Send + Sync + scale_info::TypeInfo> SignedExtension for ContractIdProvides<T>
+where
+    <T as frame_system::Config>::RuntimeCall: IsSubType<Call<T>>,
+{
+    const IDENTIFIER: &'static str = "ContractIdProvides";
+    type AccountId = T::AccountId;
+    type Call = T::RuntimeCall;
+    type AdditionalSigned = ();
+    type Pre = ();
+
+    fn additional_signed(&self) -> Result<(), TransactionValidityError> {
+        Ok(())
+    }
+
+    fn validate(
+        &self,
+        _who: &Self::AccountId,
+        call: &Self::Call,
+        _info: &DispatchInfoOf<Self::Call>,
+        _len: usize,
+    ) -> TransactionValidity {
+        if let Some(local_call) = call.is_sub_type() {
+            if let Call::bill_contract_for_block { contract_id } = local_call {
+                return ValidTransaction::with_tag_prefix(Self::IDENTIFIER)
+                    .and_provides(contract_id.to_le_bytes().to_vec())
+                    .build()
+                    .into();
+            }
+        }
+        Ok(ValidTransaction::default())
+    }
+
+    fn pre_dispatch(
+        self,
+        who: &Self::AccountId,
+        call: &Self::Call,
+        info: &DispatchInfoOf<Self::Call>,
+        len: usize,
+    ) -> Result<Self::Pre, TransactionValidityError> {
+        self.validate(who, call, info, len).map(|_| ())
+    }
+}
+
+impl<T: Config + Send + Sync + scale_info::TypeInfo> Debug for ContractIdProvides<T>
+where
+    <T as frame_system::Config>::RuntimeCall: IsSubType<Call<T>>,
+{
+    #[cfg(feature = "std")]
+    fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+        write!(f, "ContractIdProvides")
+    }
+
+    #[cfg(not(feature = "std"))]
+    fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+        Ok(())
+    }
+}
+
+impl<T: Config + Send + Sync + scale_info::TypeInfo> Default for ContractIdProvides<T>
+where
+    <T as frame_system::Config>::RuntimeCall: IsSubType<Call<T>>,
+{
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T: Config + Send + Sync + scale_info::TypeInfo> ContractIdProvides<T>
+where
+    <T as frame_system::Config>::RuntimeCall: IsSubType<Call<T>>,
+{
+    pub fn new() -> Self {
+        Self(PhantomData)
+    }
+}
```

Comment:
> Do you have any related documentation to be able to understand how this implementation do what you are expecting?
Not well documented, but the usage of [ValidTransaction's](https://paritytech.github.io/polkadot-sdk/master/sp_runtime/transaction_validity/struct.ValidTransaction.html#) provided tags inspired by the offchain-worker example here https://github.com/paritytech/polkadot-sdk/blob/e4f8a6de49b37af3c58d9414ba81f7002c911551/substrate/frame/examples/offchain-worker/src/lib.rs#L709-L714

The example uses an unsigned extrinsic. Therefore, I had to adjust for our use case and implement a [signed extension](https://paritytech.github.io/polkadot-sdk/master/sp_runtime/traits/trait.SignedExtension.html) - ([docs](https://docs.substrate.io/reference/transaction-format/)).

Reply:
> If I understand it allows to get rid of the is_next_block_author() mechanism?

Yes, the function `is_next_block_author()` was utilized to determine the suitable timing to send a transaction without incurring fees. If the transaction authored the block, it would pay the fees for itself. However, this wasn't working correctly because the off-chain worker is not part of the block production process, and you can't rely solely on timing the submission.

This was substituted with on-chain verification, where any validators (not only the block author) are allowed to submit this externsic free of charge.
https://github.com/threefoldtech/tfchain/blob/38842d6d3942807ff9435f7971eff08c629d70ec/substrate-node/pallets/pallet-smart-contract/src/lib.rs#L555-L568

Now all the validators would submit the same transactions, which allows for redundancy and a stronger guarantee that any contract will get billed at least once.

But we don't want to waste the network resources by including the same transaction multiple times, right?
This is where the signed extension comes into play, which allows the tx pool to reject the duplicated transactions more natively.